opal/mca/accelerator/base/accelerator_base_select.c

```diff
@@ -63,83 +52,32 @@ int opal_accelerator_base_select(void)
-        if (NULL == aoc) {
-            return OPAL_ERR_OUT_OF_RESOURCE;
-        }
-        OBJ_CONSTRUCT(aoc, opal_list_item_t);
-        aoc->accelerator_component = component;
-        opal_list_append(&opened, (opal_list_item_t*) aoc);
-    }
-
-    /* Finished querying all components. Check for no components found */
 
-    if( NULL == best_component ) {
-        return OPAL_ERR_NOT_FOUND;
-    }
-
-    opal_output_verbose(10, opal_accelerator_base_framework.framework_output,
-                         "selected %s best priority %d\n",
-                         best_component->base_version.mca_component_name, best_priority);
-
-    /* Save the winner */
-    accelerator_base_selected_component = *best_component;
-    opal_accelerator = *best_module;
-
-    /* Finalize all non-selected components */
-
-    for (item = opal_list_remove_first(&opened);
-         NULL != item;
-         item = opal_list_remove_first(&opened)) {
-        aoc = (accelerator_opened_component_t *) item;
-
-        if (aoc->accelerator_component != best_component ) {
-            /* Finalize */
-
-            if (NULL != aoc->accelerator_component->accelerator_finalize) {
-
-                /* Blatently ignore the return code (what would we do to
-                   recover, anyway?  This component is going away, so errors
-                   don't matter anymore) */
-
-                aoc->accelerator_component->accelerator_finalize();
-                opal_output_verbose(10, opal_accelerator_base_framework.framework_output,
-                                    "select: component %s not selected / finalized",
-                                    aoc->accelerator_component->base_version.mca_component_name);
-            }
+        module = component->accelerator_init();
+        if (NULL != module && !found_component) {
+            found_component = true;
+            accelerator_base_selected_component = *component;
+            opal_accelerator = *module;
         }
-        OBJ_DESTRUCT(aoc);
-        free(aoc);
+        else if (NULL != module && found_component) {
+            opal_output(0,
+                        "Accelerator framework only supports one type of component but both %s and %s initialized",
+                         accelerator_base_selected_component.base_version.mca_component_name,
+                         component->base_version.mca_component_name);
+            module = NULL;
+            break;
+        } /* Else case you can't initialize a component */
     }
-    OBJ_DESTRUCT(&opened);
 
-    /* This base function closes, unloads, and removes from the
-       available list all unselected components.  The available list will
-       contain only the selected component. */
-
-    mca_base_components_close(opal_accelerator_base_framework.framework_output,
-                              &opal_accelerator_base_framework.framework_components,
-                              (mca_base_component_t *) best_component);
+    if (NULL == module) {
```

Comment:
you shouldn't need this.  In fact, it's wrong.  The NULL component is just another component; it's in the list above.  Thinking about it more (sorry, I missed this when we were talking yesterday), if you're going to have a NULL component, you need to use priorities and handle selecting the first one that succeeds and ordering and all that (which kind of sucks), because otherwise you'll always have two components if there is an accelerator (the one that works and the one that is the NULL component).  You could probably explicitly check for that case instead (and skip the NULL component initialization), but you somehowe need to handle the fact that the NULL component is in the list above.

Reply:
My thought was that I didn't want to register the NULL component so that this logic would be correct.

Reply:
just confirmed... the implementation only goes around the OPAL_LIST_FOREACH once, the NULL Component isn't registered in ` &opal_accelerator_base_framework.framework_components` variable

Reply:
then something is broken; it certainly should be in there.

Reply:
broken as intended :)... will investigate.